### PR TITLE
add `max_sequential` functionality

### LINKED ISSUE #24: Add `max_sequential` functionality
We need to add the [`max_sequential` functionality](https://github.com/ANNUBS/annubes/blob/34d9787ba81ac0a4e949aab539444b7e887857a1/annubes/task.py#L85) to the class/module we will develop.

### PR DESCRIPTION
**Pull request by gcroci2**

### CODE REVIEW

**Comment by gcroci2** [comment]
* **#67** <a href="https://app.graphite.dev/github/pr/ANNUBS/neurogym/67?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/ANNUBS/neurogym/67?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#65** <a href="https://app.graphite.dev/github/pr/ANNUBS/neurogym/65?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#63** <a href="https://app.graphite.dev/github/pr/ANNUBS/neurogym/63?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: neurogym/envs/annubes.py [hunk 1]
```diff
@@ -24,6 +24,8 @@ class AnnubesEnv(TrialEnv):
             Defaults to 1000.
         catch_prob: Probability of catch trials in the session. Must be between 0 and 1 (inclusive).
             Defaults to 0.5.
+        max_sequential: It sets the maximum number of sequential trials of the same modality.
+            Defaults to None (no maximum).
```
Comment (DaniBodor) [inline neurogym/envs/annubes.py]:
```suggestion
        max_sequential: Maximum number of sequential trials of the same modality.
            Defaults to None (no maximum).
```

File: neurogym/envs/annubes.py [hunk 2]
```diff
@@ -136,7 +142,20 @@ def _new_trial(self) -> dict:
...
+                # Force a different modality
+                available_modalities = [mod for mod in self.session if mod != self.last_modality]
+                stim_type = self._rng.choice(available_modalities)
+            else:
+                stim_type = self._rng.choice(list(self.session.keys()), p=list(self.session.values()))
+                # Update sequential count
+
+            self.sequential_count = 1 if stim_type != self.last_modality else self.sequential_count + 1
+            self.last_modality = stim_type
+
```
Comment (DaniBodor) [inline neurogym/envs/annubes.py:158]: As coded, the `max_sequential` will not be applied to catches, meaning there could be many catch trials in a row. I don't think that that is the intent for this setting.

In fact, even for the case of a single non-catch modality, you would still want this setting to apply to not have too many of either the catch or the non-catch situation.

File: tests/test_annubes_env.py [hunk 3]
```diff
@@ -100,7 +106,91 @@ def test_fix_time_types(time, expected_type):
...
+        trial_types.append(env.trial["stim_type"])
+
+    # Check for sequences longer than max_sequential, excluding None (catch trials)
+    for i in range(len(trial_types) - max_sequential):
+        sequence = [t for t in trial_types[i : i + max_sequential + 1] if t is not None]
+        if len(sequence) > max_sequential:
+            assert len(set(sequence)) > 1, f"Found a sequence longer than {max_sequential} at index {i}"
+
+    # Check that all the trial types occur
+    assert set(trial_types) - {None} == set(env.session.keys()), "Not all trial types occurred"
```
Comment (DaniBodor) [inline tests/test_annubes_env.py:172]: There is no code enforcing this in the module. There could very well be a situation where there is VXVXVXVXVXVVXVXVXVXVX, in an {A,V,X} session, without throwing any error. We should either enforce that that doesn't happen in the core code, or remove this test.

File: tests/test_annubes_env.py [hunk 4]
```diff
@@ -100,7 +106,91 @@ def test_fix_time_types(time, expected_type):
...
+    trial_types: list[str | None] = []
+    for _ in range(N_TRIALS):
+        env.new_trial()
+        trial_types.append(env.trial["stim_type"])
+
+    # Check for sequences longer than max_sequential, excluding None (catch trials)
+    for i in range(len(trial_types) - max_sequential):
+        sequence = [t for t in trial_types[i : i + max_sequential + 1] if t is not None]
+        if len(sequence) > max_sequential:
+            assert len(set(sequence)) > 1, f"Found a sequence longer than {max_sequential} at index {i}"
```
Comment (DaniBodor) [inline tests/test_annubes_env.py:169]: This looks right to me (apart from the fact that i do think we should test for catches as well, see comment above). 
However, I remember in the previous repo we also had a similar looking test that ultimately turned out not to be accurate. Have you looked back at that one to see whether the mistake we made there was avoided here? You could also copy the test I wrote in the other repo to be sure that it works as expected.

File: tests/test_annubes_env.py [hunk 5]
```diff
@@ -100,7 +106,91 @@ def test_fix_time_types(time, expected_type):
...
+    # Check that the distribution is roughly balanced
+    expected_non_catch = N_TRIALS * (1 - catch_prob)
+
+    for trial_type, expected_prob in session.items():
+        count = trial_types.count(trial_type)
+        expected_count = expected_non_catch * expected_prob
+        # Allow for 20% deviation from expected count
+        assert (
+            0.8 * expected_count <= count <= 1.2 * expected_count
+        ), f"{trial_type} trials are not balanced. Expected: {expected_count:.2f}, Actual: {count}"
```
Comment (DaniBodor) [inline tests/test_annubes_env.py]: as for the "all types exist" test, there is no code actually enforcing this. I know we had this test also in the previous repo, but even there I think it was kinda iffy (I ended up decreasing the tolerance further and further to avoid it failing, which kind of defeated the purpose). 
I think we should either enforce that there is a good distribution in the code, as we do for avoiding max sequential, or not test for it at all.

File: tests/test_annubes_env.py [hunk 6]
```diff
@@ -100,7 +106,91 @@ def test_fix_time_types(time, expected_type):
...
+
+    # Check catch trial frequency
+    catch_count = trial_types.count(None)
+    expected_catch = N_TRIALS * catch_prob
+    assert (
+        0.8 * expected_catch <= catch_count <= 1.2 * expected_catch
+    ), f"Catch trials are not balanced. Expected: {expected_catch:.2f}, Actual: {catch_count}"
+
```
Comment (DaniBodor) [inline tests/test_annubes_env.py]: idem above

File: neurogym/envs/annubes.py (hunk 2 above)
Comment (gcroci2) [reply neurogym/envs/annubes.py:158]: > As coded, the `max_sequential` will not be applied to catches, meaning there could be many catch trials in a row. I don't think that that is the intent for this setting.

The limitation is meant for only the modalities, not on the catches. In case of a single non-catch modality, `max_sequential` does not apply. This was also how it was originally designed in the old annubes (it may have changed when you picked up the implementation).

File: tests/test_annubes_env.py (hunk 3 above)
Comment (gcroci2) [reply tests/test_annubes_env.py:172]: I think that for the realistic cases instead this is probabilistically enforced. There might be edge cases in which the probability of a modality is so low that it is never chosen in `stim_type = self._rng.choice(list(self.session.keys()), p=list(self.session.values()))`. In that test we are testing realistic cases, so I think it should stay even if indeed it is not enforced in all edge cases. If in the future we realize that edge cases are actually relevant we will enforce it in the code.

File: tests/test_annubes_env.py (hunk 4 above)
Comment (gcroci2) [reply tests/test_annubes_env.py:169]: I double checked this one and it should work as expected, but in case you see a bug there let me know :)

File: tests/test_annubes_env.py (hunk 5 above)
Comment (gcroci2) [reply tests/test_annubes_env.py]: As above, if we are testing realistic cases I think having such a test makes sense.

File: neurogym/envs/annubes.py (hunk 2 above)
Comment (DaniBodor) [reply neurogym/envs/annubes.py:158]: Indeed, at least in my implementation in Annubes, it did apply also to catches. What is the argument to not apply the limitation to catch trials? When thinking about it from the experimentalists point of view, I believe that they would apply the same limitations to a sequence of catch trials as to a non-catch trials.
I think the confusion arises because the order in which we implement things: first we pick a modality, and next we pick whether or not it converts to a catch. In the lab, a catch would be just one of the options (in addition to any of the modalities) for what the next trial could be.

File: tests/test_annubes_env.py (hunk 5 above)
Comment (DaniBodor) [reply tests/test_annubes_env.py]: I am responding here for both this comment and the [one above](https://github.com/ANNUBS/neurogym/pull/67#discussion_r1846427952), because my concern is greater here than above (but it does apply to both):

OK, I see the point of having these tests. If either fails, there is likely something off in the code, which causes one of the trial types not to be hit. However, it is still possible that it is just a random occurrence, right? In fact, if I turn off the random seeding and loop the test 10 times, it never passes. I've run it maybe a dozen times now and it hasn't passed all 10 cycles once.

So let's say that a developer makes some changes that create the random seed to input a different sequence and now the test fails, what message does this send to the developer? They will likely start searching their code for bugs, which maybe don't exist.
Also, they may make some other changes in the meantime that don't necessarily touch this piece but of code and now the test passes again. Personally, I would be quite frustrated by this, because in the back of my mind, there would be some bug, which I now don't know what it is or how to reproduce.

If we want to keep these tests, maybe the error/fail message needs to be better documented, e.g. something like stating that a different seed should be tested before looking for bugs.
Specifically for this test, I think a 20% deviation is too small. Even at 30% deviation allowed the 3rd parameter set fails about 5-15% of the time. (I've put the script to test that onto [this temporary branch](https://github.com/ANNUBS/neurogym/blob/test_max_sequential/tests/test_annubes_env.py))

Comment (DaniBodor) [inline tests/test_annubes_env.py]: this is an int (number expected).

```suggestion
        ), f"{trial_type} trials are not balanced. Expected: {expected_count}, Actual: {count}"
```

File: tests/test_annubes_env.py [hunk 7]
```diff
@@ -100,7 +106,91 @@ def test_fix_time_types(time, expected_type):
...
+        assert (
+            0.8 * expected_count <= count <= 1.2 * expected_count
+        ), f"{trial_type} trials are not balanced. Expected: {expected_count:.2f}, Actual: {count}"
+
+    # Check catch trial frequency
+    catch_count = trial_types.count(None)
+    expected_catch = N_TRIALS * catch_prob
+    assert (
+        0.8 * expected_catch <= catch_count <= 1.2 * expected_catch
+    ), f"Catch trials are not balanced. Expected: {expected_catch:.2f}, Actual: {catch_count}"
```
Comment (DaniBodor) [inline tests/test_annubes_env.py]: also an int

```suggestion
    ), f"Catch trials are not balanced. Expected: {expected_catch}, Actual: {catch_count}"
```

File: neurogym/envs/annubes.py (hunk 2 above)
Comment (gcroci2) [reply neurogym/envs/annubes.py:158]: I'm pretty sure catch was not seen as a modality in this regard, but we can double check

Comment (gcroci2) [reply neurogym/envs/annubes.py:158]: Added as a point to the next meeting agenda

File: tests/test_annubes_env.py (hunk 5 above)
Comment (gcroci2) [reply tests/test_annubes_env.py]: I see your point, I'd say for now let's increase the tolerance to 30%. Doing that on `test_annubes_env_max_sequential` using 42, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9 random seeds with all of the three tests listed there gives no errors to me. I think for now this is good enough, but feel free to open an issue about it for enforcing this more strictly within the code.

Comment (gcroci2) [reply tests/test_annubes_env.py]: Will be taken care in #70

**Review by DaniBodor — APPROVED** [review]
